File: OpenDiablo2.Common/Models/BitMuncher.cs
using System;
using System.IO;

namespace OpenDiablo2.Common.Models
{
    public sealed class BitMuncher
    {
        private readonly byte[] data;
        
        public int Offset { get; private set; }
        public int BitsRead { get; private set; }

        public BitMuncher(byte[] data, int offset = 0)
        {
            this.data = data;
            this.Offset = offset;
        }

        public BitMuncher(BitMuncher source)
        {
            this.data = source.data;
            this.Offset = source.Offset;
        }
        public UInt32 GetBit()
        {
            var result = (UInt32)(data[Offset / 8] >> (Offset % 8)) & 0x01;
            Offset++;
            BitsRead++;
            return result;
        }

        public void SkipBits(int bits) => Offset += bits;

        public byte GetByte() => (byte)GetBits(8);
        public Int32 GetInt32() => MakeSigned(GetBits(32), 32);
        public UInt32 GetUInt32() => GetBits(32);

        public UInt32 GetBits(int bits)
        {
            if (bits == 0)
                return 0;

            var result = 0U;
            for (var i = 0; i < bits; i++)
                result |= (GetBit() << i);

            return result;
        }

        public int GetSignedBits(int bits) => MakeSigned(GetBits(bits), bits);


        private Int32 MakeSigned(UInt32 value, int bits)
        {
            if (bits == 0)
                return 0;
            // If its a single bit, a value of 1 is -1 automagically
            if (bits == 1)
                return -(Int32)value;

            // If there is no sign bit, return the value as is
            if ((value & (1 << (bits - 1))) == 0)
                return (Int32)value;

            // We need to extend the signed bit out so that the negative value
            // representation still works with the 2s compliment rule.
            var result = UInt32.MaxValue;
            for (byte i = 0; i < bits; i++)
            {
                if (((value >> i) & 1) == 0)
                    result -= (UInt32)(1 << i);
            }

            return (Int32)result; // Force casting to a signed value
        }

    }
}

File: d2common/BitMuncher.go
package d2common

type BitMuncher struct {
	data     []byte
	Offset   int
	BitsRead int
}

func CreateBitMuncher(data []byte, offset int) *BitMuncher {
	return &BitMuncher{
		data:     data,
		Offset:   offset,
		BitsRead: 0,
	}
}

func CopyBitMuncher(source *BitMuncher) *BitMuncher {
	return &BitMuncher{
		source.data,
		source.Offset,
		0,
	}
}

func (v *BitMuncher) GetBit() uint32 {
	result := uint32(v.data[v.Offset/8]>>uint(v.Offset%8)) & 0x01
	v.Offset++
	v.BitsRead++
	return result
}

func (v *BitMuncher) SkipBits(bits int) {
	v.Offset += bits
	v.BitsRead += bits
}

func (v *BitMuncher) GetByte() byte {
	return byte(v.GetBits(8))
}

func (v *BitMuncher) GetInt32() int32 {
	return v.MakeSigned(v.GetBits(32), 32)
}

func (v *BitMuncher) GetUInt32() uint32 {
	return v.GetBits(32)
}

func (v *BitMuncher) GetBits(bits int) uint32 {
	if bits == 0 {
		return 0
	}
	result := uint32(0)
	for i := 0; i < bits; i++ {
		result |= v.GetBit() << uint(i)
	}
	return result
}

func (v *BitMuncher) GetSignedBits(bits int) int {
	return int(v.MakeSigned(v.GetBits(bits), bits))
}

func (v *BitMuncher) MakeSigned(value uint32, bits int) int32 {
	if bits == 0 {
		return 0
	}
	// If its a single bit, a value of 1 is -1 automagically
	if bits == 1 {
		return -int32(value)
	}
	// If there is no sign bit, return the value as is
	if (value & (1 << uint(bits-1))) == 0 {
		return int32(value)
	}
	// We need to extend the signed bit out so that the negative value  representation still works with the 2s compliment rule.
	result := uint32(4294967295)
	for i := byte(0); i < byte(bits); i++ {
		if ((value >> uint(i)) & 1) == 0 {
			result -= uint32(1 << uint(i))
		}
	}
	// Force casting to a signed value
	return int32(result)
}
